cpp/src/arrow/flight/transport/grpc/grpc_client.cc

```diff
@@ -702,10 +828,22 @@ class GrpcClientImpl : public internal::ClientTransport {
     stub_ = pb::FlightService::NewStub(
         ::grpc::experimental::CreateCustomChannelWithInterceptors(
             grpc_uri.str(), creds, args, std::move(interceptors)));
+
+#ifdef GRPC_ENABLE_ASYNC
+    garbage_bin_ = std::make_shared<GrpcGarbageBin>();
+#endif
+
     return Status::OK();
   }
 
   Status Close() override {
+#ifdef GRPC_ENABLE_ASYNC
+    // XXX: if there are async RPCs running when the client is
+    // stopped, then when they go to use the garbage bin, they'll
+    // instead synchronously dispose of resources from the callback
+    // thread, and will likely crash.
```

Comment:
Perhaps open an issue for this?

Reply:
I don't think we can handle this any other way, other than by punting to a global queue or similar. Something has to own the resources.

Reply:
The other way would be to track and wait for all RPCs to finish (adding the overhead on every call, and possibly defeating the point of this work in the first place.)

Reply:
I referenced the existing issue in the source.